### PR TITLE
WIP: support cockroachdb as a database

### PR DESCRIPTION
**Pull request by techknowlogick**
In most cases cockroachdb is a drop-in replacement for postgres, hopefully we can ensure support for it as well.

### CODE REVIEW

**Comment by techknowlogick** [comment]
I've just setup cockroachdb locally, and was able to install gitea, however some pages errord out, for example clicking on both the "issues", and "pull request" links in global navigation.

Below you can see an example query that failed.
```
SELECT "repository"."id"
FROM "repository"
INNER JOIN "team_user" ON repository.owner_id = team_user.org_id
INNER JOIN "team_repo" ON ($1 != $2 and repository.is_private != $3) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)
WHERE (team_user.uid = $4)
GROUP BY repository.id
Data: [true false true 563404340263223297]
routers/user/home.go:427:Issues() [E] ctxUser.GetAccessRepoIDs: pq: could not determine data type of placeholder $1
```

Running it with the data in the query itself succeeds, see below:
```
SELECT "repository"."id" FROM "repository" INNER JOIN "team_user" ON repository.owner_id = team_user.org_id INNER JOIN "team_repo" ON (true != false and repository.is_private != true) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) WHERE (team_user.uid = 563404340263223297) GROUP BY repository.id;
  id
------
(0 rows)

Time: 7.461ms

```

**Comment by techknowlogick** [comment]
When I take out injecting `true` into the query it succeeds, see:
```
Join("INNER", "team_repo", "(true != ? and repository.is_private != ?) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)", u.IsRestricted, true).
```
from models/user.go#L676

**Comment by techknowlogick** [comment]
` to_timestamp()` in heatmap query may need to be updated too.

File: models/user.go [hunk 1]
```diff
@@ -673,7 +673,7 @@ func (u *User) GetOrgRepositoryIDs(units ...UnitType) ([]int64, error) {
 	if err := x.Table("repository").
 		Cols("repository.id").
 		Join("INNER", "team_user", "repository.owner_id = team_user.org_id").
-		Join("INNER", "team_repo", "(? != ? and repository.is_private != ?) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)", true, u.IsRestricted, true).
+		Join("INNER", "team_repo", "(true != ? and repository.is_private != ?) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)", u.IsRestricted, true).
```
Comment (techknowlogick) [inline models/user.go:676]: fails on mssql :(

**Comment by lunny** [comment]
> When I take out injecting `true` into the query it succeeds, see:
> 
> ```
> Join("INNER", "team_repo", "(true != ? and repository.is_private != ?) OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)", u.IsRestricted, true).
> ```
> 
> from models/user.go#L676

So this should be a feature cockroachdb hasn't supported but postgres did.

File: models/user.go (hunk 1 above)
Comment (lunny) [reply models/user.go:676]: Some databases use 0 or 1 but not false or true. that's why we use a placeholder.

**Comment by techknowlogick** [comment]
> So this should be a feature cockroachdb hasn't supported but postgres did.

Likely. I'm going to look more into differences so I can create small test cases where it fails so I can report to cockroach.

**Comment by 6543** [comment]
I think support for cockroachDB should go into xorm first  - this should handle things like https://github.com/go-gitea/gitea/pull/11875#discussion_r439696664 too

**Comment by kolaente** [comment]
@6543 Per xorm's docs, cockroach is supported: https://gobook.io/read/gitea.com/xorm/manual-en-US/

**Comment by lunny** [comment]
@6543 xorm has supported cockroachdb, but we used some original SQLs which is not supported by cockroachdb.

**Comment by 6543** [comment]
@lunny https://xorm.io/ should be updated then - "Databases supported" doesn't list it

**Comment by kolaente** [comment]
On the other hand, cockroach claims to be 100% postgres compatible, so maybe this should be raised to them since the query seems to work on postgres

**Comment by 6543** [comment]
Fair point! @kolaente did you create an issue on there repo?

**Comment by kolaente** [comment]
@6543 Not yet, will do this week.

**Pull request closed without merging**